src/server/auth/server/saml.go

```diff
@@ -0,0 +1,430 @@
+package server
+
+import (
+	"encoding/json"
+	"encoding/xml"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path"
+	"time"
+
+	"github.com/crewjam/saml"
+	logrus "github.com/sirupsen/logrus"
+
+	"github.com/pachyderm/pachyderm/src/client/auth"
+	"github.com/pachyderm/pachyderm/src/server/pkg/backoff"
+)
+
+// canonicalConfig contains the values specified in an auth.AuthConfig proto
+// message, but as structured Go types. This is populated and returned by
+// validateConfig
+type canonicalConfig struct {
+	Version int64
+
+	ACSURL      *url.URL
+	MetadataURL *url.URL
+	DashURL     *url.URL
+
+	IDPName        string
+	IDPDescription string
+	IDPMetadataURL *url.URL
+	IDPMetadata    *saml.EntityDescriptor
+}
+
+func (c *canonicalConfig) ToProto() *auth.AuthConfig {
```

Comment:
Is it possible to have this struct just be a proto and thus obviate the need for the `ToProto` method, or would that be too inconvenient?

Reply:
The intention here is that it sort of is, in the form of AuthConfig—canonicalConfig is an AuthConfig that's already been validated, and the urls are all converted to *url.URL and such. ToProto just converts back to an AuthConfig.